design/src/rfcs/rfc0041_identity_cache_partitions.md

```diff
@@ -0,0 +1,293 @@
+
+c1.operation1()
+    .customize()
+    .config_override(config_override);
+    .send()
+    .await;
+
+c1.operation2()
+    .customize()
+    .config_override(config_override);
+    .send()
+    .await;
+```
+
+By default if an identity resolver does not provide it's own cache partition then `operation1` and `operation2` will
+be wrapped in new `SharedIdentityResolver` instances and get distinct cache partitions. If `my_custom_provider()` 
+provides it's own cache partition then `operation2` will see the cached results. 
+
+Users can control this by wrapping their provider into a `SharedCredentialsProvider` which will claim it's own
+cache partition.
+
+```rust,ignore
+
+let scoped_creds = SharedCredentialsProvider::new(my_custom_provider());
+let config_override = c1
+        .config()
+        .to_builder()
+        .set_credentials_provider(Some(scoped_creds));
+...
+```
+
+
+How to actually implement this RFC
+----------------------------------
+
+In order to implement this RFC implementations of `ResolveIdentity` need to be allowed to provide their own cache
+partition.
+
+```rust,ignore
+pub trait ResolveIdentity: Send + Sync + Debug {
+    ...
+
+    /// Returns the identity cache partition associated with this identity resolver.
+    ///
+    /// By default this returns `None` and cache partitioning is left up to `SharedIdentityResolver`.
+    /// If sharing instances of this type should use the same partition then you should override this
+    /// method and return a claimed partition.
+    fn cache_partition(&self) -> Option<IdentityCachePartition> {
+        None
+    }
+
+}
+```
+
+Crucially cache partitions must remain globally unique so this method returns `IdentityCachePartition` which is
+unique by construction. It doesn't matter if partitions are claimed early by an implementation of `ResolveIdentity`
+or at the time they are wrapped in `SharedIdentityResolver`.
+
+This is because `SdkConfig` stores instances of `SharedCredentialsProvider` (or `SharedTokenProvider`) rather than
+`SharedIdentityResolver` which is what currently knows about cache partitioning. By allowing implementations of `ResolveIdentity`
+to provide their own partition then `SharedCredentialsProvider` can claim a partition at construction time
+and return that which will re-use the same partition anywhere that the provider is shared.
+
+
+```rust,ignore
+#[derive(Clone, Debug)]
+pub struct SharedCredentialsProvider(Arc<dyn ProvideCredentials>, IdentityCachePartition);
+
+impl SharedCredentialsProvider {
+    pub fn new(provider: impl ProvideCredentials + 'static) -> Self {
+        Self(Arc::new(provider), IdentityCachePartition::new())
+    }
+}
+
+impl ResolveIdentity for SharedCredentialsProvider {
+    ...
+
+    fn cache_partition(&self) -> Option<IdentityCachePartition> {
+        Some(self.1)
+    }
+}
+
+```
+
+Additionally a new behavior version must be introduced that conditionally creates a default `IdentityCache` on `SdkConfig`
```

Comment:
Most of the defaults are provided by runtime plugins later on. Why would this one get a default at this stage?

Reply:
If we don't then each client gets their own identity cache rather than sharing one. Even if we fix the cache partitioning the default behavior will still end up being that multiple clients created from `SdkConfig` don't actually share cached results.

This _can_ be worked around of course by explicitly providing a cache to `SdkConfig` but if the default behavior we _want_ is to enable caching it probably makes sense to just make it so.